tapchannel/aux_sweeper.go

```diff
@@ -0,0 +1,402 @@
+		allocs = append(allocs, &Allocation{
+			Type:         CommitAllocationToLocal,
+			Amount:       localAsset.Amount.Val,
+			AssetVersion: asset.V1,
+			BtcAmount:    tapsend.DummyAmtSats,
+			ScriptKey:    scriptKey,
+			SortTaprootKeyBytes: schnorr.SerializePubKey(
+				scriptKey.PubKey,
+			),
+		})
+	}
+
+	// With the allocations created above, we'll now extract a slice of
+	// each of the inputproofs for each asset.
+	inputProofs := fn.Map(sweepInputs, func(o *cmsg.AssetOutput) *proof.Proof {
+		return &o.Proof.Val
+	})
+
+	// With the proofs constructed, we can now distribute the coins to
+	// create the vPackets that we'll pass on to the next stage.
+	vPackets, err := DistributeCoins(
+		inputProofs, allocs, &a.cfg.ChainParams,
+	)
+	if err != nil {
+		return lfn.Errf[[]*tappsbt.VPacket]("error distributing "+
+			"coins: %w", err)
+	}
+
+	// Next, we'll prepare all the vPackets for the sweep transaction.
+	for idx := range vPackets {
+		err := tapsend.PrepareOutputAssets(ctx, vPackets[idx])
+		if err != nil {
+			return lfn.Errf[[]*tappsbt.VPacket]("unable to "+
+				"prepare output assets: %w", err)
+		}
+	}
+
+	return lfn.Ok(vPackets)
+}
+
+// signSweepVpackets...
+func (a *AuxSweeper) signSweepVpackets(vPackets []*tappsbt.VPacket,
+	signDesc input.SignDescriptor, scriptTree input.TapscriptDescriptor,
+	ctrlBlkBytes []byte) error {
+
+	// Before we sign below, we also need to generate the tapscript With
+	// the vPackets prepared, we can now sign the output asset we'll create
+	// at a later step.
+	for _, vPacket := range vPackets {
+		// Each vPacket only has a single input, as we're sweeping a
+		// single asset from our commitment output.
+		vIn := vPacket.Inputs[0]
+
+		// In this case, the witness isn't special, so we'll set the
+		// control block now for it.
+		vIn.TaprootLeafScript[0].ControlBlock = ctrlBlkBytes
+
+		signingKey, leafToSign := applySignDescToVIn(
+			signDesc, vIn, &a.cfg.ChainParams,
+			scriptTree.TapTweak(),
+		)
+
+		// With everything set, we can now sign the new leaf we'll
+		// sweep into.
+		signed, err := a.cfg.Signer.SignVirtualPacket(
+			vPacket, tapfreighter.SkipInputProofVerify(),
+			tapfreighter.WithValidator(&schnorrSigValidator{
+				pubKey:     signingKey,
+				tapLeaf:    lfn.Some(leafToSign),
+				signMethod: input.TaprootScriptSpendSignMethod,
+			}),
+		)
+		if err != nil {
+			return fmt.Errorf("error signing virtual "+
+				"packet: %w", err)
+		}
+
+		if len(signed) != 1 || signed[0] != 0 {
+			return fmt.Errorf("error signing virtual packet, " +
+				"kgot no sig")
+		}
+
+	}
+
+	return nil
+}
+
+type sweepPkt struct {
+	vPkts []*tappsbt.VPacket
+
+	sweepDesc tapscriptSweepDesc
+}
+
+// createAndSignSweepVpackets...
+func (a *AuxSweeper) createAndSignSweepVpackets(
+	sweepInputs []*cmsg.AssetOutput, signDesc input.SignDescriptor,
+	sweepDesc lfn.Result[tapscriptSweepDesc],
+) lfn.Result[[]*tappsbt.VPacket] {
+
+	// TODO(roasbeef): is this model actually worth it? maybe need AndThen
```

Comment:
I'd say it doesn't really help readability... At least that's my first impression :sweat_smile:

Reply:
Hehe, experimenting with some new stuff 🤓. Main thing I like here is that in cases when all you have are results, then you don't need any `if err != nil` in the function body, instead it goes only at the end. It also forces you to break up your logic into discrete computations/transformations, which are assembled together at the very end. 

Agree the syntax _does_ leave some to be desired though. 

Other benefit is that you can just accept/return `T` for generic funcs, and don't need to worry about: "oh, what if there's an error too?". I also added a way to map `(T, error)` into `Result[T]`.